docs/developer/architecture/ci.rst

```diff
@@ -0,0 +1,155 @@
+
+Pre-merge
+---------
+
+What should be tested in pre-merge on every branch used during development
+(``user/*``, ``feature/*``, ``improvement/*``, ``bugfix/*``, ``w/*``).
+The pre-merge test should not long too much time (less than 40 minutes
+is great) so we can't test everything in pre-merge, we should only test
+building of the product and check that product still usable.
+
+- Building tests
+
+  - Build
+  - Lint
+  - Unit tests
+
+- Installation tests
+
+  - Simple install RHEL
+  - Simple install CentOs + expansion
+
+When merging several pull requests at the same time, given that we are on a
+queue branch (``q/*``), we may want to have or we may require additional tests
+as combinaison of several PRs could have larger impact than all individual PR:
+
+- Simple upgrade/downgrade
+
+.. _post-merge:
+
+Post-merge
+----------
+
+On each and every ``development/2.*`` branches we want to run complex tests,
+that take more time or need more ressources than classic tests that run during
+pre-merge, to ensure that the current product continues to work well.
+
+Nightly
+^^^^^^^
+
+- Upgrade, downgrade tests:
+
+  - For previous development branch
+
+    e.g.: on ``development/2.x`` test upgrade from ``development/2.(x-1)``
+    and downgrade to ``development/2.(x-1)``
+
+    - Build branch ``development/2.(x-1)`` (or retrieve it if available)
+    - Tests:
+
+      - Single node test
+      - Complex deployment test
+
+  - For last released version of current minor
+
+    e.g.: on ``development/2.x`` when developing "2.x.y-dev" test upgrade from
+    ``metalk8s-2.x.(y-1)`` and downgrade to ``metalk8s-2.x.(y-1)``
+
+    - Single node test
+    - Complex deployment test
+
+  - For last released version of previous minor
+
+    e.g.: on ``development/2.x`` when developing "2.x.y-dev" test upgrade from
+    ``metalk8s-2.(x-1).z`` and downgrade to ``metalk8s-2.(x-1).z`` where
+    "2.(x-1).z" is the last patch released version for "2.(x-1)" (``z`` may be
+    different from ``y``)
+
+    - Single node test
+    - Complex deployment test
+
+- Backup, restore tests:
+
+  - Environment with at least 3-node etcd cluster, destroy the bootstrap node
+    and spawning a new fresh node for restoration
+  - Environment with at least 3-node etcd cluster, destroy the bootstrap node
+    and use one existing node for restoration
+
+- Solutions tests
+
+.. note::
+
+  Complex deployment is (to be validated):
+
+  - 1 bootstrap
+  - 1 etcd and control
+  - 1 etcd and control and workload
+  - 1 workload and infra
+  - 1 workload
+  - 1 infra
+
+.. todo::
+
+   - Describe solutions tests (#1993)
+
+Weekly
+^^^^^^
+
+More complex tests (TBD)
+
+Adaptive test plan
+------------------
+
+CI pre-merge may be more flexible by including some logic about the content
+of the changeset.
+
+For example:
+
+- For a commit that changes solely/uniquely documentation, we don't need/want
+  to run all the installation test but rather run tests related to
+  documentation.
```

Comment:
` For a commit that changes uniquely documentation, we don't need
  to run the entire installation test suite but rather run tests related to
  documentation.`